File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>

using namespace std;

int n, m;

int arr[30];
int res;
    
void recur(int cur, int sum){
    if(cur == n){
        if(sum == m) res++;
        return;
    }
    recur(cur+1, sum + arr[cur]);
    recur(cur+1, sum);
}
int main()
{
    cin >> n >> m;
    for(int i = 0; i < n; i++){
        cin >> arr[i];
    }
    
    recur(0, 0);
    if(m == 0) res--;
    cout << res;
    
    return 0;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.*;
import java.util.*;

public class Main {

    static int arr[] = new int[22];
    static int N, S;
    static int ret = 0;

    public static void main(String[] args) throws Exception {
        // System.setIn(new FileInputStream("src/input.txt"));
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        S = sc.nextInt();
        for (int i = 0; i < N; i++) {
            arr[i] = sc.nextInt();
        }

        recur(0, 0);

        if (S == 0) {
            ret--;
        }
        System.out.println(ret);
    }

    static void recur(int cur, int sum) {
        if (cur == N) {
            if (sum == S) {
                ret++;
            }
            return;
        }
        recur(cur + 1, sum + arr[cur]);
        recur(cur + 1, sum);
    }
}